### PR TITLE
Collect metrics & assert on solver outcomes

### PR DESCRIPTION
**Pull request by aggelgian**
Add the functionality to collect metrics in a new component (gen_server) that can be called from everywhere in the code.
We now just collect a metric with the status of every solver outcome.

The metrics can be optionally reported at the end of the executions with a flag.

```
$./cuter example foo11 '[[17]]' -r -s 8 -d 10 -m
Compiling example.erl ...
ok
Testing example:foo11/1 ...
.x
example:foo11("*")
x.
example:foo11(2)
.xxxxxxxxx...xxxxx..xxx

=== Collected Metrics ===

[solver/status]
- UNSAT: 19
- SAT: 9

=== Inputs That Lead to Runtime Errors ===
#1 example:foo11("*")
#2 example:foo11(2)
```

Furthermore, we can assert on the number of solver statuses in functional tests.
With the definition
```json
        {
            "module": "search",
            "function": "f",
            "args": "[0]",
            "depth": "25",
            "errors": true,
            "arity": 1,
            "solutions": [
                "[3]"
            ],
            "solver": {
              "SAT": 3,
              "UNSAT": 7
            },
            "skip": false
        }
```
we get the following failure message
```bash
$make ftest no=97

[Test #97]
/home/aggelos/Projects/cuter/cuter search f '[0]' -d 25 -p 2 -s 24 -pa /home/aggelos/Projects/cuter/test/ftest/ebin -m --sorted-errors
.xx
search:f(3)
.x.xxx
SOLVER MISMATCH
=================================================================
EXPECTED: {'SAT': 3, 'UNSAT': 7}
GOT: {'SAT': 3, 'UNSAT': 6}
=================================================================
make: *** [Makefile:201: ftest] Error 1
```

### CODE REVIEW

File: src/cuter.erl [hunk 1]
```diff
@@ -362,3 +381,7 @@ suppress_unsupported_mfas(Options) ->
 -spec type_normalization([option()]) -> boolean().
 type_normalization(Options) ->
   not lists:member(?NO_TYPE_NORMALIZATION, Options).
+
+-spec report_metrics([option()]) -> boolean().
+report_metrics(Options) ->
+  lists:member(?REPORT_METRICS, Options).
```
Comment (kostis) [inline src/cuter.erl:387]: `lists:member` works here, but perhaps we should consider making the `Options` a property list at some point (in a separate refactoring step) and use the `proplists` functions instead.

File: src/cuter_analyzer.erl [hunk 2]
```diff
@@ -199,3 +201,27 @@ calculate_coverage(Feasible, Visited) ->
...
+-spec report_metrics() -> ok.
+report_metrics() ->
+  case cuter_metrics:get_distribution_metrics() of
+    [] -> ok;
+    Names ->
+      cuter_pp:pp_metrics_title(),
+      report_distribution_metrics(Names)
+  end.
+
+report_distribution_metrics([]) ->
```
Comment (kostis) [inline src/cuter_analyzer.erl:218]: For uniformity with the function above, the `ok`s should not be in a separate line in this function.

File: include/cuter_metric_definitions.hrl [hunk 3]
```diff
@@ -0,0 +1,10 @@
+%% -*- erlang-indent-level: 2 -*-
```
Comment (kostis) [inline include/cuter_metric_definitions.hrl:1]: I think we do not need the `_definitions` part in the name of this header file...

File: test/ftest/src/search.erl [hunk 4]
```diff
@@ -0,0 +1,19 @@
+-module(search).
```
Comment (kostis) [inline test/ftest/src/search.erl:1]: `search` is a very general name for this test module; I suggest you change it to a more specific one.

More importantly, I suggest you add some comment describing what this test module is supposed to be testing.

File: test/utest/src/cuter_metrics_tests.erl [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+-include_lib("eunit/include/eunit.hrl").
+
+-spec test() -> 'ok' | {'error', term()}.  %% This should be provided by EUnit
+
+-spec distribution_metric_test() -> ok.
+distribution_metric_test() ->
+  Name = some_metric,
+  ok = cuter_metrics:start(),
+  ok = cuter_metrics:define_distribution_metric(Name),
+  ok = cuter_metrics:measure_distribution(Name, 42),
```
Comment (kostis) [inline test/utest/src/cuter_metrics_tests.erl:14]: These 6 lines can become just one:
```erlang
   lists:foreach(fun (N) -> cuter_metrics:measure_distribution(Name, N) end, [42, 1, 42, 23, 1, 42]),
```
A similar change can occur in the next function (with xxx, etc.) and there we can actually use the same list in both the call to `foreach` and in the `assertEqual`.

**Review by kostis — APPROVED** [review]
Looks OK, but I've left some suggestions.

File: src/cuter.erl (hunk 1 above)
Comment (aggelgian) [reply src/cuter.erl:387]: ack.

File: src/cuter_analyzer.erl (hunk 2 above)
Comment (aggelgian) [reply src/cuter_analyzer.erl]: Done..

File: test/ftest/src/search.erl (hunk 4 above)
Comment (aggelgian) [reply test/ftest/src/search.erl]: Done.

File: include/cuter_metric_definitions.hrl (hunk 3 above)
Comment (aggelgian) [reply include/cuter_metric_definitions.hrl]: Done.

File: test/utest/src/cuter_metrics_tests.erl (hunk 5 above)
Comment (aggelgian) [reply test/utest/src/cuter_metrics_tests.erl]: Done.

**Pull request merged**